contracts/origination/OriginationControllerSTIRFRY.sol

```diff
@@ -0,0 +1,268 @@
+
+        address vaultAddress = vaultFactory.instanceAt(loanTerms.collateralId);
+        _validateStirfryTerms(loanTerms, stirfryData, vaultAddress);
+
+        {
+            // signature validation
+            Side neededSide = isSelfOrApproved(borrower, msg.sender) ? Side.LEND : Side.BORROW;
+
+            address signingCounterparty = neededSide == Side.LEND ? lender : borrower;
+            address callingCounterparty = neededSide == Side.LEND ? borrower : lender;
+
+            (bytes32 sighash, address externalSigner) = _recoverSignature(loanTerms, sig, sigProperties, neededSide, signingCounterparty, itemPredicates);
+
+            _validateCounterparties(signingCounterparty, callingCounterparty, msg.sender, externalSigner, sig, sighash);
+
+            // consume signer nonce
+            loanCore.consumeNonce(externalSigner, sigProperties.nonce, sigProperties.maxUses);
+        }
+
+        // initialize loan
+        loanId = _initialize(loanTerms, stirfryData, vaultAddress, borrower, lender);
+
+        // Run predicates check at the end of the function, after vault is in escrow. This makes sure
+        // that re-entrancy was not employed to withdraw collateral after the predicates check occurs.
+        if (itemPredicates.length > 0) originationHelpers.runPredicatesCheck(borrower, lender, loanTerms, itemPredicates);
+    }
+
+    // =========================================== HELPERS ==============================================
+
+    /**
+     * @notice Validate that the stirfry data provided is valid for the given loan terms.
+     *
+     * @param loanTerms                     The terms of the loan.
+     * @param stirfryData                   The stirfry data to validate.
+     * @param vaultAddress                  The address of the vault.
+     */
+    function _validateStirfryTerms(
+        LoanLibrary.LoanTerms calldata loanTerms,
+        StirfryData calldata stirfryData,
+        address vaultAddress
+    ) internal view {
+        // verify the vaulted currency is allowed to be paired with the loan terms payable currency
+        bytes32 currencyHash = keccak256(
+            abi.encodePacked(
+                loanTerms.payableCurrency,
+                stirfryData.vaultedCurrency,
+                stirfryData.payableToVaultedCurrencyRatio
+            )
+        );
+        if(stirfryPairs[currencyHash] != true)
+            revert OCS_InvalidStirfryPair(loanTerms.payableCurrency, stirfryData.vaultedCurrency);
+
+        // verify the vaulted currency amounts
+        if(loanTerms.principal * stirfryData.payableToVaultedCurrencyRatio != stirfryData.lenderVaultedCurrencyAmount)
+            revert OCS_InvalidPrincipalAmounts(
+                loanTerms.principal,
+                stirfryData.payableToVaultedCurrencyRatio,
+                stirfryData.lenderVaultedCurrencyAmount
+            );
+
+        // verify that the vault holds the lenderVaultedCurrencyAmount
```

Comment:
Should we verify that funds are held in the vault, or should we actually collect funds from the lender and send them into the vault at this time? Thinking about the scenario where the vault is created and funded but does not instantly get filled; if the vault is pre-funded, then the borrower earns that pre-swap yield, right?

Reply:
See comments below as well; could we just collect `loanTerms.principal * stirfryData.payableToVaultedCurrencyRatio`?

Reply:
Yes, agree the OC should bundle the assets together.

When implementing this I found that we need to mint the Asset Vault to the OC before collecting the ERC20 tokens because LoanCore does not implement the IERC721Receiver interface. Which just means there is an extra transfer of the vault from the OC to LC before calling `loanCore.startLoan()`. But I also think this is safer from a security standpoint since the vault factory uses `safeMint()` and there would be a callback on LoanCore if it was minted to LC.